### ISSUE TITLE
other context info

### ISSUE THREAD

**Issue by duke-arioch** [phantamanta44/rphud#1]
I may want to extend this to provide FOV entity and looked at block information to the context. I'd be happy to fork and do pull request etc - just wondering if this is the kind of thing you would want in your mod as right now it is player and world focused instead of surroundings focused. I've done a bit of work with horses off zyin's mod and think the addition of a "relative to relevant block" widget location type would also be useful.

**Comment by phantamanta44** [comment]
@duke-arioch there are really only two restrictions I'm concerned about:

First, I've written various "server-specific client" mods for [Brawl](https://brawl.com), and this mod was designed almost expressly for the purpose of being a sort of future-proof catch-all mod for those servers. As such, I designed this mod with the rules of that  server in mind, which includes a "no damage indicators!" rule for whatever reason. Thus, any feature that makes famage indicators possible is a no-no.

Second, I want this mod to be as noob-friendly as possible -- that is, I don't want any features that would require more than basic algebra and 2-dimensional spacial reasoning to understand and use properly. So a `highlighted_block_id` sort of variable is fine, but a `raytrace_block_id(x, y, z)` function is somewhat less desirable.

**Comment by duke-arioch** [comment]
yeah here's my thoughts on the looking at (and entities in FOV) - 

First, context - I would only add simple[ish] contents to the context.

Second, there would have to be some new kind of rendering potentially for FOV or WAILA type blocks and entities - renderonblock or renderonenttity(offset x,,y)

Third, what do you think of a more css kind of display adaptation or is that already being done in another mod? dont want to make it too complex or heavy for your maintenance so might just fork for that part.

Details:

**Context**

- There are sign HUDs but meh - sign contents might be nice. 
- Horses and other friendly mobs in view might be nice to have stats for - maybe minus current HP if the server is paranoid. I have made modifications to MLK's horse stats mod that included foal/adult, tame/wild and pattern/color. 
- For blocks you could add light level. I know theres a light level mod but might be nice
- Distance to looking at block
- Biome, weather etc

**Display locations**

DisplayOnBlockFace (facing closest or directional (TOP, SOUTH, ETC)
DisplayOnEntity (Offset from entity in coords relative to entity or yourself)

**Display Widgets/content**

probably with the extra display locations you could do what was needed with existing? If not, maybe optionally CSS-styling could be used?

These are really mostly utility functions and I have no desire to turn this into a primary  pvp-focused plugin. If you disagree with the overall or parts, I can just fork this but wanted to see if parts might be interesting for you to have back in. Interested to know your thoughts and like the modular, reusable nature of your mod!

**Comment by phantamanta44** [comment]
**Context stuff:**
I don't think sign info is really appropriate for this mod. The information on a sign is already readily available, and I don't think it would be particularly desirable for that information to be more available than it is -- that's erring dangerously close to a hacked client sort of thing.

Stats for mobs sounds interesting, but only the mob currently being looked at should have its information exposed. That way, the end user isn't forced to deal with some kind of iteration or similar. I think perhaps including health is okay, but it might be a good idea to include some sort of flag the server can pass to the client (maybe via chat, like Rei's minimap?) to disable it.

I don't think light level is a good addition. It would be somewhat unclear what block the information is drawn from, and I don't think that sort of nondeterministic-ness is good.

The length of the look raytrace seems okay, but performance might become an issue if such a raytrace were being performed every render tick over a potentially long distance.

Biome and weather information is perfectly fine. I didn't include those data because they didn't pertain to Brawl gamemodes, but I suppose there might be other applications for rphud 😛 

**Display location stuff:**
I think these might be implementable as sections? Something like this:
```
-- components --
# stuff renders on HUD

-- components/entity --
# stuff renders on plane facing player near the entity

-- components/block --
# stuff renders on plane facing player near the block
```
I don't think these should be parameterized such that they can be manipulated in 3D space. At most, pack creators should only have to work in 2 dimensions. Additionally, I think the rendered HUD components should always face the player, because that's kind of the point of a HUD lol.

As for making the lang more CSS-y, I don't think that would be a good move. The mod was designed such that resource pack makers can build HUDs without needing any meaningful understanding of programming, and I feel dipping into the hierarchical world of CSS selectors and cascading styles is a little too far. Perhaps if you really wanted to, you could build a scriptable HUD mod that parsed and rendered the classic markup-styling-behaviour model, but I don't think that's the sort of direction I want rphud to go in.

**Comment by duke-arioch** [comment]
Thanks - that helps clear things up a bit. What I will look at first then is-

Biome, weather
Looked at block, 
Distance (up to some limit)
Animal info (i will leave out HP for now but you could add if you like)

Anything related to 3-D rendering or to styling - css/markdown/html/whatever - I will play with afterward in my own version and of course you can incorporate if you like.

As for signs and light level - I can leave those till later and add to my own. There are other HUDs that do that anyway!
